### PR TITLE
[HealthKit] Add support for Xcode 15 beta 7

### PR DESCRIPTION
**Pull request by haritha-mohan**
Looking into a intro failure on iOS: Selector not found for HealthKit.HKHealthStore : workoutSessionMirroringStartHandler in System.Action`1[HealthKit.HKWorkoutSession] get_WorkoutSessionMirroringStartHandler() on HealthKit.HKHealthStore

### CODE REVIEW

File: src/HealthKit/Enums.cs [hunk 1]
```diff
@@ -721,4 +723,34 @@ public enum HKVisionPrescriptionType : ulong {
...
 		Contacts,
 	}
+
+	[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0)]
+	[Native]
+	public enum HKCyclingFunctionalThresholdPowerTestType : long {
+		MaxExercise60Minute = 1,
+		MaxExercise20Minute,
+		RampTest,
+		PredictionExercise
```
Comment (mandel-macaque) [inline src/HealthKit/Enums.cs:733]:
```suggestion
		PredictionExercise,
```

File: src/HealthKit/Enums.cs [hunk 2]
```diff
@@ -721,4 +723,34 @@ public enum HKVisionPrescriptionType : ulong {
...
+	}
+
+	[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0)]
+	[Native]
+	public enum HKPhysicalEffortEstimationType : long {
+		ActivityLookup = 1,
+		DeviceSensed
```
Comment (mandel-macaque) [inline src/HealthKit/Enums.cs:740]:
```suggestion
		DeviceSensed,
```

File: src/HealthKit/Enums.cs [hunk 3]
```diff
@@ -721,4 +723,34 @@ public enum HKVisionPrescriptionType : ulong {
...
+	}
+
+	[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0)]
+	[Native]
+	public enum HKWaterSalinity : long {
+		FreshWater = 1,
+		SaltWater
```
Comment (mandel-macaque) [inline src/HealthKit/Enums.cs:747]:
```suggestion
		SaltWater,
```

File: src/HealthKit/Enums.cs [hunk 4]
```diff
@@ -721,4 +723,34 @@ public enum HKVisionPrescriptionType : ulong {
...
+	}
+
+	[Watch (10, 0), iOS (17, 0)]
+	[Native]
+	public enum HKWorkoutSessionType : long {
+		Primary = 0,
+		Mirrored
```
Comment (mandel-macaque) [inline src/HealthKit/Enums.cs:754]:
```suggestion
		Mirrored,
```

File: src/healthkit.cs [hunk 5]
```diff
@@ -717,6 +720,10 @@ interface HKHealthStore {
 
 		// HKWorkout category
 
+		[Deprecated (PlatformName.iOS, 17, 0, message: "Use HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler) instead.")]
+		[Deprecated (PlatformName.WatchOS, 10, 0, message: "Use HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler) instead.")]
+		[Deprecated (PlatformName.MacCatalyst, 16, 0, message: "Use HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler) instead.")]
+		[Deprecated (PlatformName.MacOSX, 14, 0, message: "Use HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler) instead.")]
```
Comment (mandel-macaque) [inline src/healthkit.cs:726]:
```suggestion
		[Deprecated (PlatformName.iOS, 17, 0, message: "Use 'HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler)' instead.")]
		[Deprecated (PlatformName.WatchOS, 10, 0, message: "Use 'HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler)' instead.")]
		[Deprecated (PlatformName.MacCatalyst, 16, 0, message: "Use 'HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler)' instead.")]
		[Deprecated (PlatformName.MacOSX, 14, 0, message: "Use 'HKWorkoutBuilder.Add (HKSample [] samples, HKWorkoutBuilderCompletionHandler completionHandler)' instead.")]
```

File: src/healthkit.cs [hunk 6]
```diff
@@ -3589,18 +3677,25 @@ interface HKWorkoutSessionDelegate {
...
 		void DidBeginActivity (HKWorkoutSession workoutSession, HKWorkoutConfiguration workoutConfiguration, NSDate date);
 
-		[Watch (9, 0), NoiOS, NoTV, Mac (13, 0), NoMacCatalyst]
+		[Watch (9, 0), NoTV, Mac (13, 0)]
 		[Export ("workoutSession:didEndActivityWithConfiguration:date:")]
 		void DidEndActivity (HKWorkoutSession workoutSession, HKWorkoutConfiguration workoutConfiguration, NSDate date);
+
+		[Watch (10, 0), iOS (17, 0), MacCatalyst (17, 0), NoTV, Mac (14, 0)]
+		[Export ("workoutSession:didReceiveDataFromRemoteWorkoutSession:")]
+		void DidReceiveDataFromRemoteWorkoutSession (HKWorkoutSession workoutSession, NSData [] data);
```
Comment (mandel-macaque) [inline src/healthkit.cs:3694]:
```suggestion
		void DidReceiveData (HKWorkoutSession workoutSession, NSData [] data);
```

File: src/healthkit.cs [hunk 7]
```diff
@@ -3589,18 +3677,25 @@ interface HKWorkoutSessionDelegate {
...
+
+		[Watch (10, 0), iOS (17, 0), MacCatalyst (17, 0), NoTV, Mac (14, 0)]
+		[Export ("workoutSession:didDisconnectFromRemoteDeviceWithError:")]
+		void DidDisconnectFromRemoteDeviceWithError (HKWorkoutSession workoutSession, [NullAllowed] NSError error);
```
Comment (mandel-macaque) [inline src/healthkit.cs:3698]:
```suggestion
		void DidDisconnect (HKWorkoutSession workoutSession, [NullAllowed] NSError error);
```

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

File: src/healthkit.cs [hunk 8]
```diff
@@ -3099,6 +3163,16 @@ interface HKUnit : NSCopying, NSSecureCoding {
 		[Static]
 		[Export ("degreeAngleUnit")]
 		HKUnit DegreeAngle { get; }
+
+		[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0), NoTV]
+		[Static]
+		[Export ("luxUnitWithMetricPrefix:")]
+		HKUnit LuxUnitWithMetricPrefix (HKMetricPrefix prefix);
```
Comment (dalexsoto) [inline src/healthkit.cs:3170]:
```suggestion
		HKUnit GetLuxUnit (HKMetricPrefix prefix);
```

File: src/healthkit.cs [hunk 9]
```diff
@@ -3099,6 +3163,16 @@ interface HKUnit : NSCopying, NSSecureCoding {
...
+
+		[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0), NoTV]
+		[Static]
+		[Export ("luxUnit")]
+		HKUnit LuxUnit ();
```
Comment (dalexsoto) [inline src/healthkit.cs:3175]:
```suggestion
		HKUnit GetLuxUnit ();
```

File: src/healthkit.cs [hunk 10]
```diff
@@ -3513,71 +3585,87 @@ interface HKWorkoutSession : NSSecureCoding {
...
+		HKWorkoutSessionType Type { get; }
+
+		[Watch (10, 0), NoTV, NoMacCatalyst, NoMac, iOS (17, 0)]
+		[Export ("sendDataToRemoteWorkoutSession:completion:")]
+		[Async]
+		void SendDataToRemoteWorkoutSession (NSData data, Action<bool, NSError> completion);
+
+		[Watch (10, 0), NoTV, NoMacCatalyst, NoMac, NoiOS]
+		[Export ("startMirroringToCompanionDeviceWithCompletion:")]
+		void StartMirroringToCompanionDeviceWithCompletion (Action<bool, NSError> completion);
```
Comment (dalexsoto) [inline src/healthkit.cs:3659]:
```suggestion
		void StartMirroringToCompanionDevice (Action<bool, NSError> completion);
```

File: src/healthkit.cs [hunk 11]
```diff
@@ -3513,71 +3585,87 @@ interface HKWorkoutSession : NSSecureCoding {
...
+
+		[Watch (10, 0), NoTV, NoMacCatalyst, NoMac, NoiOS]
+		[Export ("stopMirroringToCompanionDeviceWithCompletion:")]
+		void StopMirroringToCompanionDeviceWithCompletion (Action<bool, NSError> completion);
```
Comment (dalexsoto) [inline src/healthkit.cs:3663]:
```suggestion
		void StopMirroringToCompanionDevice (Action<bool, NSError> completion);
```

**Review by dalexsoto — CHANGES_REQUESTED** [review]
👍

**Review by mandel-macaque — APPROVED** [review]

File: src/healthkit.cs [hunk 12]
```diff
@@ -3099,6 +3163,16 @@ interface HKUnit : NSCopying, NSSecureCoding {
...
+
+		[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0), NoTV]
+		[Static]
+		[Export ("luxUnitWithMetricPrefix:")]
+		HKUnit GetLuxUnit (HKMetricPrefix prefix);
+
+		[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0), NoTV]
+		[Static]
+		[Export ("luxUnit")]
+		HKUnit GetLuxUnit ();
```
Comment (rolfbjarne) [inline src/healthkit.cs:3175]: We've bound other "*Unit" API without the "Unit" suffix in this type, and as a property:

```suggestion
		HKUnit Lux { get; }
```

this also matches the name in Swift.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/healthkit.cs [hunk 13]
```diff
@@ -3513,71 +3585,87 @@ interface HKWorkoutSession : NSSecureCoding {
...
+		HKWorkoutSessionType Type { get; }
+
+		[Watch (10, 0), NoTV, NoMacCatalyst, NoMac, iOS (17, 0)]
+		[Export ("sendDataToRemoteWorkoutSession:completion:")]
+		[Async]
+		void SendDataToRemoteWorkoutSession (NSData data, Action<bool, NSError> completion);
+
+		[Watch (10, 0), NoTV, NoMacCatalyst, NoMac, NoiOS]
+		[Export ("startMirroringToCompanionDeviceWithCompletion:")]
+		void StartMirroringToCompanionDevice (Action<bool, NSError> completion);
```
Comment (tj-devel709) [inline src/healthkit.cs:3659]: I think this method should have the [Async] attribute similar to the method you added above
The thing that triggered me to check is the completion Action parameter.

I confirmed with the web docs [here](https://developer.apple.com/documentation/healthkit/hkworkoutsession/4165515-startmirroringtocompaniondevicew?language=objc) and you can scroll down to see the photo below where they call it in the sample code with the await keyword
image

File: src/healthkit.cs [hunk 14]
```diff
@@ -3513,71 +3585,87 @@ interface HKWorkoutSession : NSSecureCoding {
...
+
+		[Watch (10, 0), NoTV, NoMacCatalyst, NoMac, NoiOS]
+		[Export ("stopMirroringToCompanionDeviceWithCompletion:")]
+		void StopMirroringToCompanionDevice (Action<bool, NSError> completion);
```
Comment (tj-devel709) [inline src/healthkit.cs:3663]: Same with this one ^

**Review by tj-devel709 — CHANGES_REQUESTED** [review]
Looks good! I think you may need to add a few Async calls though!

File: src/healthkit.cs [hunk 15]
```diff
@@ -3099,6 +3163,16 @@ interface HKUnit : NSCopying, NSSecureCoding {
 		[Static]
 		[Export ("degreeAngleUnit")]
 		HKUnit DegreeAngle { get; }
+
+		[Watch (10, 0), MacCatalyst (17, 0), Mac (14, 0), iOS (17, 0), NoTV]
+		[Static]
+		[Export ("luxUnitWithMetricPrefix:")]
+		HKUnit CreateLuxUnit (HKMetricPrefix prefix);
```
Comment (rolfbjarne) [inline src/healthkit.cs:3170]: Sorry, I got confused, _this_ matches (most of) the other methods in this class:

```suggestion
		HKUnit CreateLux (HKMetricPrefix prefix);
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/healthkit.cs (hunk 15 above)
Comment (haritha-mohan) [reply src/healthkit.cs]: @rolfbjarne ah yep I missed that too, fixed

File: src/healthkit.cs (hunk 13 above)
Comment (haritha-mohan) [reply src/healthkit.cs]: @tj-devel709 yep that's what intro was complaining about as well, fixed!

**Review by tj-devel709 — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**